Support wildcard within sync_list entries

abraunegg opened the pull request:
* Support wildcard within sync_list entries

abraunegg commented:
**Config:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --display-config
Configuration file successfully loaded
onedrive version                       = v2.4.5-12-g0c770ef
Config path                            = /home/alex/.config/onedrive-personal/
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /home/alex/OneDrivePersonal
Config option 'skip_dir'               = 
Config option 'skip_file'              = ~*|.~*|*.tmp|*.swp
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 30
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = true
sync_list contents:
# Include all files and folders under /random_files/
random_files/temp*
# Include all files and folders under /random_images/
random_images/
Business Shared Folders configured     = false
```
Folder composition:
```
.
├── Documents
│   ├── Images
│   └── Music
├── Dropbox
├── Images
├── Music
├── random_files -> /home/alex/OneDrive/random_files
│   ├── blah
│   │   └── Ge2YFs6HDUsEmwpbA8b80RsV8APZZk6l
│   ├── temp
│   ├── temp1
│   │   └── 3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY
│   └── temp2
│       ├── 9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
│       └── EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
├── random_images -> /home/alex/OneDrive/random_images
│   ├── 5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN
│   ├── eyy9NflXfPMxleNYnN4DTanVm3WV6m3p
│   ├── TCdX2javImiZnd5A3FyaqOrkXoIOI9if
│   └── UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3
└── random_videos -> /home/alex/OneDrive/random_videos
    ├── hM0IGhCMbrhDKTUqjxBCi0ryAmECjMNI
    └── y7Q5Qw7RHQU1VbOq0pCZIPZwv8GRaAjj

23 directories
```


Expectation is that `random_files/temp*` is included

**Without PR:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose --resync
Using 'user' Config Dir: /home/alex/.config/onedrive-personal/
Using 'system' Config Dir: 
Configuration file successfully loaded
Deleting the saved status ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.4.5-12-g0c770ef
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5334903675
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 123 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents
Skipping item - excluded by sync_list config: random_videos
Skipping item - excluded by sync_list config: Dropbox
Skipping item - excluded by sync_list config: Images
Skipping item - excluded by sync_list config: Music
Skipping item - excluded by sync_list config: file1.txt
Skipping item - excluded by sync_list config: random_files/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
Skipping item - excluded by sync_list config: random_files/Ge2YFs6HDUsEmwpbA8b80RsV8APZZk6l
Skipping item - excluded by sync_list config: random_files/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
Skipping item - excluded by sync_list config: random_files/temp1
Skipping item - excluded by sync_list config: random_files/temp2
Skipping item - excluded by sync_list config: random_files/temp/file0.data
Skipping item - excluded by sync_list config: random_files/temp/file1.data
Skipping item - excluded by sync_list config: random_files/temp/file2.data
Skipping item - excluded by sync_list config: random_files/temp/file3.data
Skipping item - excluded by sync_list config: random_files/temp/file4.data
Skipping item - excluded by sync_list config: random_files/temp/file5.data
Skipping item - excluded by sync_list config: random_files/temp/file6.data
Skipping item - excluded by sync_list config: random_files/temp/file7.data
Skipping item - excluded by sync_list config: random_files/temp/file8.data
Skipping item - excluded by sync_list config: random_files/temp/file9.data
Number of items from OneDrive to process: 1
Uploading differences of ~/OneDrivePersonal
Processing .
The directory has not changed
Processing random_images
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image0.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image1.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image2.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image3.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image5.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image4.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image8.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image9.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image6.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image7.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3
The directory has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image0.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image1.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image2.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image4.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image6.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image3.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image8.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image5.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image9.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN
The directory has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image0.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image2.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image1.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image3.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image6.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image8.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image5.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image4.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if
The directory has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image1.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image3.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image5.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image4.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image7.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image2.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image6.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image0.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image8.png
The file has not changed
Processing random_files
The directory has not changed
Processing random_files/temp
The directory has not changed
Uploading new items of ~/OneDrivePersonal
Skipping item - excluded by sync_list config: ./random_files/temp1
Skipping item - excluded by sync_list config: ./random_files/temp2
Skipping item - excluded by sync_list config: ./random_files/blah
Skipping item - excluded by sync_list config: ./random_files/temp/file0.data
Skipping item - excluded by sync_list config: ./random_files/temp/file1.data
Skipping item - excluded by sync_list config: ./random_files/temp/file2.data
Skipping item - excluded by sync_list config: ./random_files/temp/file3.data
Skipping item - excluded by sync_list config: ./random_files/temp/file4.data
Skipping item - excluded by sync_list config: ./random_files/temp/file5.data
Skipping item - excluded by sync_list config: ./random_files/temp/file6.data
Skipping item - excluded by sync_list config: ./random_files/temp/file7.data
Skipping item - excluded by sync_list config: ./random_files/temp/file8.data
Skipping item - excluded by sync_list config: ./random_files/temp/file9.data
Skipping item - excluded by sync_list config: ./random_videos
Skipping item - excluded by sync_list config: ./Dropbox
Skipping item - excluded by sync_list config: ./Documents
Skipping item - excluded by sync_list config: ./Images
Skipping item - excluded by sync_list config: ./Music
Skipping item - excluded by skip_file config: ./file1.tmp
Skipping item - excluded by sync_list config: ./file1.txt
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 1 OneDrive items to ensure consistent local state due to a full scan being requested
```

Result is that `random_files/temp*` is excluded


**With PR:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose --resync
Using 'user' Config Dir: /home/alex/.config/onedrive-personal/
Using 'system' Config Dir: 
Configuration file successfully loaded
Deleting the saved status ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.4.5-12-g0c770ef
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5334903675
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 123 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents
Skipping item - excluded by sync_list config: random_videos
Skipping item - excluded by sync_list config: Dropbox
Skipping item - excluded by sync_list config: Images
Skipping item - excluded by sync_list config: Music
Skipping item - excluded by sync_list config: file1.txt
Skipping item - excluded by sync_list config: random_files/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
Skipping item - excluded by sync_list config: random_files/Ge2YFs6HDUsEmwpbA8b80RsV8APZZk6l
Skipping item - excluded by sync_list config: random_files/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
Number of items from OneDrive to process: 1
Uploading differences of ~/OneDrivePersonal
Processing .
The directory has not changed
Processing random_images
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image0.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image1.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image2.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image3.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image5.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image4.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image8.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image9.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image6.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image7.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3
The directory has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image0.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image1.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image2.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image4.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image6.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image3.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image8.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image5.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image9.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN
The directory has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image0.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image2.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image1.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image3.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image6.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image8.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image5.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image4.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if
The directory has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image1.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image3.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image5.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image4.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image7.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image2.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image6.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image0.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image8.png
The file has not changed
Processing random_files
The directory has not changed
Processing random_files/temp
The directory has not changed
Processing random_files/temp/file0.data
The file has not changed
Processing random_files/temp/file1.data
The file has not changed
Processing random_files/temp/file2.data
The file has not changed
Processing random_files/temp/file3.data
The file has not changed
Processing random_files/temp/file4.data
The file has not changed
Processing random_files/temp/file5.data
The file has not changed
Processing random_files/temp/file6.data
The file has not changed
Processing random_files/temp/file7.data
The file has not changed
Processing random_files/temp/file8.data
The file has not changed
Processing random_files/temp/file9.data
The file has not changed
Processing random_files/temp1
The directory has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY
The directory has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file0.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file1.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file2.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file3.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file4.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file5.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file6.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file7.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file8.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file9.data
The file has not changed
Processing random_files/temp2
The directory has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
The directory has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file0.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file1.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file2.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file3.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file4.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file5.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file6.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file7.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file8.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file9.data
The file has not changed
Processing random_files/temp2/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
The directory has not changed
Processing random_files/temp2/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo/file4.data
The file has not changed
Uploading new items of ~/OneDrivePersonal
Skipping item - excluded by sync_list config: ./random_files/blah
Skipping item - excluded by sync_list config: ./random_videos
Skipping item - excluded by sync_list config: ./Dropbox
Skipping item - excluded by sync_list config: ./Documents
Skipping item - excluded by sync_list config: ./Images
Skipping item - excluded by sync_list config: ./Music
Skipping item - excluded by skip_file config: ./file1.tmp
Skipping item - excluded by sync_list config: ./file1.txt
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 1 OneDrive items to ensure consistent local state due to a full scan being requested
```

Result is that `random_files/temp*` is included

norbusan commented on src/selective.d:
```diff
@@ -202,15 +204,33 @@ private bool isPathExcluded(string path, string[] allowedPaths)
 			return false;
 		}
+		// is the allowed path terminated by a wildcard? (*)
+		if (allowed.endsWith("*")) {
```
That gives at least a partial answer for that, but `*` isn't allowed in any other location, right? OTOH, we have 
```
Config option 'skip_file'              = ~*|.~*|*.tmp|*.swp
```
so `*` are evaluated there also in other positions.

I fear that people will use synclist entries like
```
random_files/img-201*-*-*/raw/
```
(just a dummy example) and expect it to work.

I am not sure what a good solution is. I see a few options, either something like the `.gitignore` configuration style, or some `find` syntax style.

But well, at least this is a good step, albeit we need to make clear that the star only works at the end of the entry, I guess.

abraunegg replied:
Yes.. the current documentation only shows '*' at end, could updated it to make it super clear that this is the only option, or play around with a search if '*' is present, then do a regex to search & match ... would be computationally expensive depending on how many '*' are being pattern matched.

Will look at it for a while and see what I can come up with.

abraunegg replied:
Should be resolved by https://github.com/abraunegg/onedrive/pull/1063/commits/191571e5c15f916aeb7ddaf3c148d589957eb844

norbusan reviewed (approved):
Approved as this is fine as is, but see my comment.

abraunegg commented:
Updated PR.

**Config:**
```
Configuration file successfully loaded
onedrive version                       = v2.4.5-12-g0c770ef
Config path                            = /home/alex/.config/onedrive-personal/
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /home/alex/OneDrivePersonal
Config option 'skip_dir'               = 
Config option 'skip_file'              = ~*|.~*|*.tmp|*.swp
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 30
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = true
sync_list contents:
# Include all files and folders under /random_files/
random_files/temp*
# Include all files and folders under /random_images/
random_ima*es/
Business Shared Folders configured     = false
```

**Result:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --resync --verbose
Using 'user' Config Dir: /home/alex/.config/onedrive-personal/
Using 'system' Config Dir: 
Configuration file successfully loaded
Deleting the saved status ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.4.5-12-g0c770ef
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5334903675
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 123 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents
Skipping item - excluded by sync_list config: random_videos
Skipping item - excluded by sync_list config: Dropbox
Skipping item - excluded by sync_list config: Images
Skipping item - excluded by sync_list config: Music
Skipping item - excluded by sync_list config: file1.txt
Skipping item - excluded by sync_list config: random_files/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
Skipping item - excluded by sync_list config: random_files/Ge2YFs6HDUsEmwpbA8b80RsV8APZZk6l
Skipping item - excluded by sync_list config: random_files/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
Number of items from OneDrive to process: 1
Uploading differences of ~/OneDrivePersonal
Processing .
The directory has not changed
Processing random_images
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p
The directory has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image0.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image1.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image2.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image3.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image5.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image4.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image8.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image9.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image6.png
The file has not changed
Processing random_images/eyy9NflXfPMxleNYnN4DTanVm3WV6m3p/image7.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3
The directory has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image0.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image1.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image2.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image4.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image6.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image3.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image8.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image5.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image9.png
The file has not changed
Processing random_images/UTv2BnXNJbZKSrUNjlK7eKVODYl7vuZ3/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN
The directory has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image0.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image2.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image1.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image3.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image6.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image7.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image8.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image5.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image4.png
The file has not changed
Processing random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if
The directory has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image1.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image3.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image5.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image4.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image7.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image9.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image2.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image6.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image0.png
The file has not changed
Processing random_images/TCdX2javImiZnd5A3FyaqOrkXoIOI9if/image8.png
The file has not changed
Processing random_files
The directory has not changed
Processing random_files/temp
The directory has not changed
Processing random_files/temp/file0.data
The file has not changed
Processing random_files/temp/file1.data
The file has not changed
Processing random_files/temp/file2.data
The file has not changed
Processing random_files/temp/file3.data
The file has not changed
Processing random_files/temp/file4.data
The file has not changed
Processing random_files/temp/file5.data
The file has not changed
Processing random_files/temp/file6.data
The file has not changed
Processing random_files/temp/file7.data
The file has not changed
Processing random_files/temp/file8.data
The file has not changed
Processing random_files/temp/file9.data
The file has not changed
Processing random_files/temp1
The directory has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY
The directory has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file0.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file1.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file2.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file3.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file4.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file5.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file6.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file7.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file8.data
The file has not changed
Processing random_files/temp1/3hzby5qSP5e47Ke2JG5obCYdnUkmJvDY/file9.data
The file has not changed
Processing random_files/temp2
The directory has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l
The directory has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file0.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file1.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file2.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file3.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file4.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file5.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file6.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file7.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file8.data
The file has not changed
Processing random_files/temp2/9kkDiggwzfimGzIC3zEMHCEIQZz43E0l/file9.data
The file has not changed
Processing random_files/temp2/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo
The directory has not changed
Processing random_files/temp2/EKP9AGkht4XU6jFs9k8yWuesv0mCOmpo/file4.data
The file has not changed
Uploading new items of ~/OneDrivePersonal
Skipping item - excluded by sync_list config: ./random_files/blah
Skipping item - excluded by sync_list config: ./random_videos
Skipping item - excluded by sync_list config: ./Dropbox
Skipping item - excluded by sync_list config: ./Documents
Skipping item - excluded by sync_list config: ./Images
Skipping item - excluded by sync_list config: ./Music
Skipping item - excluded by skip_file config: ./file1.tmp
Skipping item - excluded by sync_list config: ./file1.txt
Applying changes of Path ID: 66D53BE8A5056ECA!101
Updated Remaining Free Space: 5334903675
Processing 1 OneDrive items to ensure consistent local state due to a full scan being requested
```

* anything `random_files/temp*` is synced
* anything `random_ima*es` is synced

norbusan commented on src/selective.d:
```diff
@@ -194,23 +197,46 @@ final class SelectiveSync
+		if (canFind(allowed, wildcard)) {
+			// allowed path contains a wildcard
+			// manually escape '/'
+			string escapedAllowed = replace(allowed, "/", "\\/");
```
Why is this necessary? `/` doesn't have a special meaning in regular expressions.
And what about `\*` as part of the file name? (but ok, that we can ignore - file names with literal `*` in it are a bit a pain)

abraunegg replied:
It needs to be escaped, so '/' in the allowed path, needs to be escaped with '\/' when passed in as a regex, thus needs to be '\\/' so that the regex can be passed correctly.

I wish I had all the development debug logs in the console, but from https://regex101.com/ - see below:

![regex1](https://user-images.githubusercontent.com/4956234/93302016-9f25c400-f83c-11ea-8331-d8794ac5eede.JPG)


![regex2](https://user-images.githubusercontent.com/4956234/93302028-a351e180-f83c-11ea-93d6-e34f76dd58f5.JPG)


![regex3](https://user-images.githubusercontent.com/4956234/93302033-a5b43b80-f83c-11ea-9cef-0329f262a48a.JPG)

norbusan replied:
Well, that is what the `sed` requires because there 's/needle/replacement/options' is the normal syntax. But I don't see that D is using this. Have you actually tried without escapting the `/`? If that is necessary, the documentation of D regex library needs to be improved ;-)

norbusan replied:
Anyway, this is in fact irrelevant, escaping is never a problem, so I approve.

abraunegg replied:
> Well, that is what the `sed` requires because there 's/needle/replacement/options' is the normal syntax. But I don't see that D is using this. Have you actually tried without escapting the `/`? If that is necessary, the documentation of D regex library needs to be improved ;-)

Yep .. the documentation needs to be improved .. the clue for me was here: https://dlang.org/phobos/std_regex.html#escaper

norbusan replied:
Hmm that doesn't say anything about `/` though?

abraunegg replied:
In testing, I tested using 'escaper' to see what it would provide, and based on the output, use that to manually manipulate the input path to ensure that there was no breakage.

I have tested this further, and it appears that the escaping `/` is not needed:
```
Evaluation against 'sync_list' for: random_images/5vuzmRS6M6QsyXEGxze8oCoIxZAciXYN/image9.png
Evaluation against 'sync_list' entry: random_files/temp*
escapedString via escaper(allowed): random_files\/temp\*
escapedAllowed via manual replace: random_files/temp.*
Evaluation against 'sync_list' entry: random_ima*es
escapedString via escaper(allowed): random_ima\*es
escapedAllowed via manual replace: random_ima.*es
Evaluation against 'sync_list' result - wildcard pattern match
```
![regex4](https://user-images.githubusercontent.com/4956234/93387599-d764eb00-f8ac-11ea-92bc-9a0134afb15a.JPG)

Will fix this up in the PR I am developing for #1039

norbusan replied:
This is what I expected. I don't know escaper, but regex ;-) Thanks, and as I said, it does not hurt so it is not a problem anyway.
